---
layout: page
title: Software
---

<!--## Contents
{:.no_toc}

* Will be replaced with the ToC, excluding the "Contents" header
{:toc}-->

## [funtooNorm](https://github.com/GreenwoodLab/funtooNorm)

> The [R](http://cran.r-project.org/) package [funtooNorm](https://github.com/GreenwoodLab/funtooNorm) provides a function for normalization of Illumina Infinium Human Methylation 450 BeadChip (Illumina 450K) data when there are samples from multiple tissues or cell types.

----

## [pcev](https://github.com/GreenwoodLab/pcev)

> [pcev](https://github.com/GreenwoodLab/pcev) is an [R](http://cran.r-project.org/) package which implements Principal components of explained variance (PCEV); it is a statistical tool which be used for the analysis of a mutivariate response vector. It is a dimension-reduction technique, similar to Principal components Analysis (PCA), which seeks to maximize the proportion of variance (in the response vector) being explained by a set of covariates. Stable release on [CRAN](https://cran.r-project.org/web/packages/pcev/index.html)

----

## [RVPedigree](https://github.com/GreenwoodLab/RVPedigree)

> [RVPedigree](https://github.com/GreenwoodLab/RVPedigree) is an [R](http://cran.r-project.org/) package which is a collection of the five region-based rare-variant genetic association tests. The following tests are currently implemented: ASKAT, ASKAT-Normalized, VC-C1, VC-C2 and VC-C3.

----

## [MURAT](https://github.com/GreenwoodLab/MURAT)

> [MURAT](https://github.com/GreenwoodLab/MURAT) is an [R](http://cran.r-project.org/) package to conduct rare-variant association tests for multiple continuous traits  

----

## [Eclust](http://sahirbhatnagar.com/eclust/)

> This software is written in the open source software environment [R](http://cran.r-project.org/). It's main functionality is to fit statistical models for analyzing interactions between a high dimensional dataset (e.g. genomics, brain imaging), the environment and a response. It automatically performs variable selection and estimation, while conserving the strong heredity property, i.e., an interaction term is selected only if the main effects are in the model.

---

## [CellTypeAdjustment](https://github.com/GreenwoodLab/CellTypeAdjustment)

> R script showing examples of running methods adjusting for cell type composition in DNA methylation studies (450K). Simulated data is available as well: [https://zenodo.org/record/46746#.VtW7MmSAOko](https://zenodo.org/record/46746#.VtW7MmSAOko)

